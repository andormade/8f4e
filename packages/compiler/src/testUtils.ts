import wabt from 'wabt';

import {
	createCodeSection,
	createExportSection,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import { FunctionBody } from './wasmUtils/typeHelpers';
import { CompiledModule, TestModule } from './types';
import { WORD_LENGTH } from './consts';

import { compileModules, getInitialMemory } from '.';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: DataView, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		if (Number.isInteger(initialMemory[i])) {
			memory.setInt32(i * 4, initialMemory[i], true);
		} else {
			memory.setFloat32(i * 4, initialMemory[i], true);
		}
	}
}

export async function createTestModule(sourceCode: string): Promise<TestModule> {
	let allocatedMemoryForTestData = 0;

	const module: CompiledModule = compileModules([{ code: sourceCode.split('\n') }])[0];
	const program = createSingleFunctionWASMProgram(module.functionBody);
	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const dataView = new DataView(memoryRef.buffer);
	const memoryBuffer = new Int32Array(memoryRef.buffer);
	let instance: WebAssembly.Instance | undefined;
	let wat = '';

	try {
		const webAssemblyInstantiatedSource = await WebAssembly.instantiate(program, {
			js: {
				memory: memoryRef,
			},
		});
		instance = webAssemblyInstantiatedSource.instance;

		wat = await new Promise(resolve => {
			wabt().then(_wabt => {
				const module = _wabt.readWasm(program, {});
				resolve(module.toText({}));
			});
		});
	} catch (error) {
		console.log(error);
	}

	const reset = () => {
		setInitialMemory(dataView, module);

		// Clear the test data that was out of the module's scope.
		for (let i = module.memoryWordSize; i < module.memoryWordSize + allocatedMemoryForTestData; i++) {
			memoryBuffer[i] = 0;
		}
		allocatedMemoryForTestData = 0;
	};

	reset();

	const test = (instance?.exports.test ||
		function () {
			return;
		}) as CallableFunction;

	const memoryGet = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);

			if (!memoryItem) {
				return;
			}

			if (memoryItem.isInteger) {
				return dataView.getInt32(memoryItem.byteAddress, true);
			} else {
				return dataView.getFloat32(memoryItem.byteAddress, true);
			}
		}

		return memoryBuffer[address];
	};

	const memorySet = (address: string | number, value: number | number[]): void => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			if (typeof value === 'number') {
				if (Number.isInteger(value)) {
					dataView.setInt32(memoryItem.byteAddress, value, true);
				} else {
					dataView.setFloat32(memoryItem.byteAddress, value, true);
				}
			} else {
				for (let i = 0; i < value.length; i++) {
					if (Number.isInteger(value[i])) {
						dataView.setInt32(memoryItem.byteAddress + i * 4, value[i], true);
					} else {
						dataView.setFloat32(memoryItem.byteAddress + i * 4, value[i], true);
					}
				}
			}
			return;
		}

		if (typeof value === 'number') {
			if (Number.isInteger(value)) {
				dataView.setInt32(address * 4, value, true);
			} else {
				dataView.setFloat32(address * 4, value, true);
			}
		} else {
			for (let i = 0; i < value.length; i++) {
				if (Number.isInteger(value[i])) {
					dataView.setInt32(address * 4 + i * 4, value[i], true);
				} else {
					dataView.setFloat32(address * 4 + i * 4, value[i], true);
				}
			}
		}
	};

	const getByteAddress = (address: string | number): number | undefined => {
		if (typeof address === 'string') {
			const memoryItem = module.memoryMap.get(address);
			if (!memoryItem) {
				return;
			}

			return memoryItem.byteAddress;
		}

		return address * WORD_LENGTH;
	};

	const allocMemoryForPointer = (address: string | number): number => {
		const firstFreeAddress = module.memoryWordSize + allocatedMemoryForTestData;
		memorySet(address, firstFreeAddress * WORD_LENGTH);
		allocatedMemoryForTestData++;
		return firstFreeAddress;
	};

	// @ts-ignore
	memoryBuffer.get = memoryGet;
	// @ts-ignore
	memoryBuffer.set = memorySet;
	// @ts-ignore
	memoryBuffer.allocMemoryForPointer = allocMemoryForPointer;
	// @ts-ignore
	memoryBuffer.byteAddress = getByteAddress;

	return {
		// @ts-ignore
		memory: memoryBuffer,
		test,
		reset,
		wat,
		program,
		memoryMap: module.memoryMap,
		ast: module.ast,
	};
}
