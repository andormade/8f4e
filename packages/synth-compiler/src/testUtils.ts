import {
	createFunctionSection,
	createTypeSection,
	createFunctionType,
	createExportSection,
	createFunctionExport,
	createCodeSection,
	createImportSection,
	createMemoryImport,
	FunctionBody,
} from '@8f4e/bytecode-utils';
import { compile } from '@8f4e/module-compiler';
import { getInitialMemory, ModuleGenerator } from '.';
import { CompiledModule, MemoryMap } from './types';
import wabt from 'wabt';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

export function createSingleFunctionWASMProgram(functionBody: FunctionBody): Uint8Array {
	return Uint8Array.from([
		...HEADER,
		...VERSION,
		...createTypeSection([createFunctionType([], [])]),
		...createImportSection([createMemoryImport('js', 'memory')]),
		...createFunctionSection([0x00]),
		...createExportSection([createFunctionExport('test', 0x00)]),
		...createCodeSection([functionBody]),
	]);
}

export function setInitialMemory(memory: Int32Array, module: CompiledModule): void {
	const initialMemory = getInitialMemory(module);
	for (let i = 0; i < initialMemory.length; i++) {
		memory[i] = initialMemory[i];
	}
}

export async function createTestModule(
	moduleCreator: ModuleGenerator | string,
	initialConfig = {}
): Promise<{
	memory: Int32Array;
	test: CallableFunction;
	reset: () => void;
	wat: string;
	program: Uint8Array;
	memoryMap: MemoryMap[];
}> {
	let module: CompiledModule;

	if (typeof moduleCreator === 'function') {
		module = moduleCreator('test', { byte: nthWord => nthWord * 4, word: nthWord => nthWord }, initialConfig);
	} else {
		module = compile(moduleCreator, 'test', 0);
	}

	const program = createSingleFunctionWASMProgram(module.functionBody);

	const memoryRef = new WebAssembly.Memory({ initial: 1 });
	const memoryBuffer = new Int32Array(memoryRef.buffer);

	const { instance } = await WebAssembly.instantiate(program, {
		js: {
			memory: memoryRef,
		},
	});

	const reset = () => {
		setInitialMemory(memoryBuffer, module);
	};

	reset();

	const test = instance.exports.test as CallableFunction;

	const wat: string = await new Promise(resolve => {
		wabt().then(_wabt => {
			const module = _wabt.readWasm(program, {});
			resolve(module.toText({}));
		});
	});

	return { memory: memoryBuffer, test, reset, wat, program, memoryMap: module.memoryMap };
}
