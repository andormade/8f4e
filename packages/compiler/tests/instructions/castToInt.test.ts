import { moduleTester } from './testUtils';

moduleTester(
	'castToInt',
	`module castToInt

float input
int output

push &output
push input
castToInt
store

moduleEnd
`,
	[
		[{ input: 1.1 }, { output: 1 }],
		[{ input: -69.69 }, { output: -69 }],
		[{ input: 0.001 }, { output: 0 }],
		[{ input: 420.99 }, { output: 420 }],
	]
);
