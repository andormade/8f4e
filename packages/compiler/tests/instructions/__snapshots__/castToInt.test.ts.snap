// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`castToInt if the generated AST, WAT and memory map match the snapshot 1`] = `
Array [
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "castToInt",
      },
    ],
    "instruction": "module",
    "lineNumber": 0,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "float",
    "lineNumber": 2,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "output",
      },
    ],
    "instruction": "int",
    "lineNumber": 3,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "&output",
      },
    ],
    "instruction": "push",
    "lineNumber": 5,
  },
  Object {
    "arguments": Array [
      Object {
        "type": "identifier",
        "value": "input",
      },
    ],
    "instruction": "push",
    "lineNumber": 6,
  },
  Object {
    "arguments": Array [],
    "instruction": "castToInt",
    "lineNumber": 7,
  },
  Object {
    "arguments": Array [],
    "instruction": "store",
    "lineNumber": 8,
  },
  Object {
    "arguments": Array [],
    "instruction": "moduleEnd",
    "lineNumber": 10,
  },
]
`;

exports[`castToInt if the generated AST, WAT and memory map match the snapshot 2`] = `
"(module
  (type (;0;) (func))
  (import \\"js\\" \\"memory\\" (memory (;0;) 1))
  (func (;0;) (type 0)
    i32.const 4
    i32.const 0
    f32.load
    i32.trunc_f32_s
    i32.store)
  (export \\"test\\" (func 0)))
"
`;

exports[`castToInt if the generated AST, WAT and memory map match the snapshot 3`] = `
Map {
  "input" => Object {
    "byteAddress": 0,
    "default": 0,
    "elementWordSize": 4,
    "id": "input",
    "isInteger": false,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "numberOfElements": 1,
    "type": "float",
    "wordAlignedAddress": 0,
    "wordAlignedSize": 1,
  },
  "output" => Object {
    "byteAddress": 4,
    "default": 0,
    "elementWordSize": 4,
    "id": "output",
    "isInteger": true,
    "isPointer": false,
    "isPointingToInteger": false,
    "isPointingToPointer": false,
    "numberOfElements": 1,
    "type": "int",
    "wordAlignedAddress": 1,
    "wordAlignedSize": 1,
  },
}
`;
